### ISSUE TITLE
Does Bap support symbolic execution

### ISSUE THREAD

**Issue by anwarmamat** [BinaryAnalysisPlatform/bap#954]
When I run a Primus Taint Analysis on the following code, it shows the malloc reaches the return of the function foo, which is actually impossible. Does it mean the Primus micro execution executes every path, including the impossible path?
```
int *foo(int *a) {
  if(*a > 0){
       return NULL;
  }else{   // *a < 0
      if(*a > 6){  //impossible path
         int *ret = malloc(sizeof(int));
         return ret;
       }else{
          return NULL;
       }
  }
}
```

**Comment by ivg** [comment]
>  Does it mean the Primus micro execution executes every path, including the impossible path?

Yes and no. Primus Taint Analysis by default is using the so called promiscuous mode of execution, which involves forced execution of all linear independent paths. Therefore, infeasible paths are also taken. To summarize. Primus Taint Analysis is driven by Primus mode of execution. The default mode visits all linearly independent paths (but not all paths).

Answering your main question. 

> Does Bap support symbolic execution.

Yes and no :) There are no public/open-source/free plugins for symbolic execution known to me. The closest open source plugin that does something similar (the weakest precondition) is CBAT's [WP][1] plugin.


[1]: https://github.com/draperlaboratory/cbat_tools/tree/master/wp/plugin
